File: macos/TokendialCore/Sources/TokendialCore/I18n/Strings.swift
import Foundation

/// The shared string catalogue in docs/i18n. English is the base; a language file overrides what it
/// translates. Plurals follow CLDR categories, dates and times follow the locale. Both platforms load the same files.
public final class Strings {
    public static let languages: [(code: String, name: String)] = [
        ("en", "English"), ("en-GB", "English (UK)"), ("es", "Español"), ("fr", "Français"), ("de", "Deutsch"), ("ar", "العربية")
    ]

    public static var directory: URL?
    private static let lock = NSLock()
    private static var base = Catalog.load("en")
    private static var current = base
    private static var localeValue = Locale.current
    public static var changed: (() -> Void)?

    public static var language: String { current.language }
    public static var locale: Locale { localeValue }
    public static var rightToLeft: Bool { current.direction == "rtl" }

    /// Point the loader at the directory holding the JSON files (the app bundle's copy of docs/i18n) and reload.
    public static func load(from directory: URL, language: String? = nil) {
        self.directory = directory
        base = Catalog.load("en")
        use(language)
    }

    /// Switch language. Nil follows the system; an unknown code falls back to English.
    public static func use(_ language: String?) {
        let code = resolve(language ?? Locale.preferredLanguages.first ?? "en")
        lock.lock()
        current = code == "en" ? base : Catalog.load(code)
        localeValue = Locale(identifier: code == "en" ? "en_US" : code == "ar" ? "ar_SA" : code.replacingOccurrences(of: "-", with: "_"))
        lock.unlock()
        changed?()
    }

    public static func resolve(_ name: String) -> String {
        if let exact = languages.first(where: { $0.code.lowercased() == name.lowercased() }) { return exact.code }
        let two = String(name.split(separator: "-").first ?? "").lowercased()
        return languages.contains { $0.code == two } ? two : "en"
    }

    public static func t(_ key: String, _ args: [String: Any] = [:]) -> String {
        fill(lookup(key) ?? key, args)
    }

    /// A plural form for n, chosen by the language's CLDR rules; {n} is filled with the number.
    public static func plural(_ key: String, _ n: Int, _ args: [String: Any] = [:]) -> String {
        var merged = args
        merged["n"] = n
        guard let forms = lookupForms(key) else { return t(key, merged) }
        let category = pluralCategory(current.language, n)
        return fill(forms[category] ?? forms["other"] ?? key, merged)
    }

    /// A provider-produced English window label in the current language; unknown labels pass through.
    public static func label(_ english: String) -> String {
        if let exact = current.labels[english] ?? base.labels[english] { return exact }
        for (pattern, key) in labelPatterns {
            guard let match = pattern.firstMatch(in: english, range: NSRange(english.startIndex..., in: english)),
                  let range = Range(match.range(at: 1), in: english),
                  let template = current.labels[key] ?? base.labels[key] else { continue }
            return template.replacingOccurrences(of: "{n}", with: String(english[range]))
        }
        return english
    }

    private static let labelPatterns: [(NSRegularExpression, String)] = [
        (try! NSRegularExpression(pattern: "^(\\d+)h limit$"), "{n}h limit"),
        (try! NSRegularExpression(pattern: "^(\\d+)m limit$"), "{n}m limit"),
        (try! NSRegularExpression(pattern: "^(\\d+)d limit$"), "{n}d limit"),
        (try! NSRegularExpression(pattern: "^Usage \\((\\d+) h\\)$"), "Usage ({n} h)"),
        (try! NSRegularExpression(pattern: "^Usage \\((\\d+) wk\\)$"), "Usage ({n} wk)")
    ]

    public static func pluralCategory(_ language: String, _ n: Int) -> String {
        switch String(language.split(separator: "-").first ?? "") {
        case "ar":
            if n == 0 { return "zero" }
            if n == 1 { return "one" }
            if n == 2 { return "two" }
            let mod = n % 100
            if (3...10).contains(mod) { return "few" }
            if (11...99).contains(mod) { return "many" }
            return "other"
        case "fr": return n == 0 || n == 1 ? "one" : "other"
        default: return n == 1 ? "one" : "other"
        }
    }

    private static func lookup(_ key: String) -> String? {
        if let s = current.strings[key] as? String { return s }
        return base.strings[key] as? String
    }

    private static func lookupForms(_ key: String) -> [String: String]? {
        (current.strings[key] as? [String: String]) ?? (base.strings[key] as? [String: String])
    }

    private static func fill(_ text: String, _ args: [String: Any]) -> String {
        var out = text
        for (name, value) in args { out = out.replacingOccurrences(of: "{\(name)}", with: "\(value)") }
        return out
    }

    public static func keys(_ language: String) -> [String: Any] { Catalog.load(language).strings }
    public static func labels(_ language: String) -> [String: String] { Catalog.load(language).labels }

    private struct Catalog {
        var language = "en"
        var direction = "ltr"
        var strings: [String: Any] = [:]
        var labels: [String: String] = [:]

        static func load(_ code: String) -> Catalog {
            guard let directory = Strings.directory, let data = try? Data(contentsOf: directory.appendingPathComponent("\(code).json")), let root = JSON.object(data) else {
                return code == "en" ? Catalog() : load("en")
            }
            var catalog = Catalog(language: root.str("language") ?? code, direction: root.str("direction") ?? "ltr")
            catalog.strings = root.obj("strings") ?? [:]
            catalog.labels = (root.obj("labels") ?? [:]).compactMapValues { $0 as? String }
            return catalog
        }
    }
}

File: windows/Tokendial.Core/I18n/Strings.cs
using System.Globalization;
using System.Reflection;
using System.Text.Json;
using System.Text.RegularExpressions;

namespace Tokendial.Core.I18n;

/// <summary>
/// The shared string catalogue in docs/i18n, embedded in the core. English is the base;
/// a language file overrides what it translates. Plurals follow CLDR categories, dates
/// and times follow the culture. Both platforms load the same files.
/// </summary>
public static class Strings
{
    public static readonly IReadOnlyList<(string Code, string Name)> Languages =
    [
        ("en", "English"), ("en-GB", "English (UK)"), ("es", "Español"), ("fr", "Français"), ("de", "Deutsch"), ("ar", "العربية")
    ];

    private static readonly object gate = new();
    private static Catalog baseCatalog = Catalog.Load("en");
    private static Catalog current = baseCatalog;
    private static CultureInfo culture = CultureInfo.CurrentUICulture;
    private static string? requested;

    public static event Action? Changed;

    /// <summary>The language in use: a code from <see cref="Languages"/>.</summary>
    public static string Language => current.Language;
    public static CultureInfo Culture => culture;
    public static bool RightToLeft => current.Direction == "rtl";

    /// <summary>Switch language. Null follows the system; an unknown code falls back to English.</summary>
    public static void Use(string? language)
    {
        var code = Resolve(language ?? CultureInfo.CurrentUICulture.Name);
        lock (gate)
        {
            requested = language;
            current = code == "en" ? baseCatalog : Catalog.Load(code);
            culture = CultureInfo.GetCultureInfo(code == "en" ? "en-US" : code == "ar" ? "ar-SA" : code);
        }
        Changed?.Invoke();
    }

    public static string? Requested => requested;

    /// <summary>"es-MX" → "es"; "en-GB" stays; anything unknown → "en".</summary>
    public static string Resolve(string name)
    {
        if (Languages.Any(l => l.Code.Equals(name, StringComparison.OrdinalIgnoreCase))) return Languages.First(l => l.Code.Equals(name, StringComparison.OrdinalIgnoreCase)).Code;
        var two = name.Split('-')[0].ToLowerInvariant();
        return Languages.Any(l => l.Code == two) ? two : "en";
    }

    public static string T(string key, params (string Name, object? Value)[] args) => Fill(Lookup(key) is string s ? s : key, args);

    /// <summary>A plural form for n, chosen by the language's CLDR rules; {n} is filled with the number.</summary>
    public static string Plural(string key, int n, params (string Name, object? Value)[] args)
    {
        var forms = LookupForms(key);
        if (forms is null) return T(key, [.. args, ("n", n)]);
        var category = PluralCategory(current.Language, n);
        var text = forms.GetValueOrDefault(category) ?? forms.GetValueOrDefault("other") ?? key;
        return Fill(text, [.. args, ("n", n)]);
    }

    /// <summary>A provider-produced English window label in the current language; unknown labels pass through.</summary>
    public static string Label(string english)
    {
        var exact = current.Labels.GetValueOrDefault(english) ?? baseCatalog.Labels.GetValueOrDefault(english);
        if (exact is not null) return exact;
        foreach (var (pattern, key) in LabelPatterns)
        {
            var m = pattern.Match(english);
            if (!m.Success) continue;
            var template = current.Labels.GetValueOrDefault(key) ?? baseCatalog.Labels.GetValueOrDefault(key);
            if (template is not null) return template.Replace("{n}", m.Groups[1].Value);
        }
        return english;
    }

    private static readonly (Regex, string)[] LabelPatterns =
    [
        (new Regex(@"^(\d+)h limit$"), "{n}h limit"),
        (new Regex(@"^(\d+)m limit$"), "{n}m limit"),
        (new Regex(@"^(\d+)d limit$"), "{n}d limit"),
        (new Regex(@"^Usage \((\d+) h\)$"), "Usage ({n} h)"),
        (new Regex(@"^Usage \((\d+) wk\)$"), "Usage ({n} wk)")
    ];

    public static string PluralCategory(string language, int n)
    {
        var lang = language.Split('-')[0];
        switch (lang)
        {
            case "ar":
                if (n == 0) return "zero";
                if (n == 1) return "one";
                if (n == 2) return "two";
                var mod = n % 100;
                if (mod is >= 3 and <= 10) return "few";
                if (mod is >= 11 and <= 99) return "many";
                return "other";
            case "fr":
                return n is 0 or 1 ? "one" : "other";
            default:
                return n == 1 ? "one" : "other";
        }
    }

    private static string? Lookup(string key)
    {
        if (current.Strings.TryGetValue(key, out var value) && value.ValueKind == JsonValueKind.String) return value.GetString();
        if (baseCatalog.Strings.TryGetValue(key, out var fallback) && fallback.ValueKind == JsonValueKind.String) return fallback.GetString();
        return null;
    }

    private static Dictionary<string, string>? LookupForms(string key)
    {
        if (!(current.Strings.TryGetValue(key, out var forms) && forms.ValueKind == JsonValueKind.Object)
            && !(baseCatalog.Strings.TryGetValue(key, out forms) && forms.ValueKind == JsonValueKind.Object)) return null;
        return forms.EnumerateObject().ToDictionary(p => p.Name, p => p.Value.GetString() ?? "");
    }

    private static string Fill(string text, (string Name, object? Value)[] args)
    {
        foreach (var (name, value) in args) text = text.Replace("{" + name + "}", Convert.ToString(value, CultureInfo.InvariantCulture) ?? "");
        return text;
    }

    /// <summary>Every key in a catalogue, for the conformance test that keeps languages aligned.</summary>
    public static IReadOnlyDictionary<string, JsonElement> Keys(string language) => Catalog.Load(language).Strings;
    public static IReadOnlyDictionary<string, string> Labels(string language) => Catalog.Load(language).Labels;

    private sealed class Catalog
    {
        public string Language = "en";
        public string Direction = "ltr";
        public Dictionary<string, JsonElement> Strings = new();
        public Dictionary<string, string> Labels = new();

        public static Catalog Load(string code)
        {
            using var stream = Assembly.GetExecutingAssembly().GetManifestResourceStream($"i18n/{code}.json");
            if (stream is null) return code == "en" ? new Catalog() : Load("en");
            using var document = JsonDocument.Parse(stream);
            var root = document.RootElement;
            var catalog = new Catalog { Language = root.GetProperty("language").GetString() ?? code, Direction = root.TryGetProperty("direction", out var d) ? d.GetString() ?? "ltr" : "ltr" };
            if (root.TryGetProperty("strings", out var strings))
                foreach (var p in strings.EnumerateObject()) catalog.Strings[p.Name] = p.Value.Clone();
            if (root.TryGetProperty("labels", out var labels))
                foreach (var p in labels.EnumerateObject()) catalog.Labels[p.Name] = p.Value.GetString() ?? p.Name;
            return catalog;
        }
    }
}
